### ISSUE TITLE
Embed in frames does only work once, having the same URI

### ISSUE THREAD

**Issue by dr0i** [jsonld-java/jsonld-java#150]
From https://github.com/json-ld/json-ld.org/issues/119 for examples and background:

> The current behaviour avoids embedding the same data multiple times in the result makes it difficult  to work with the output. A proposal to change this to "agressive re-embedding" is currently being  discussed.

Using my own words:
Declaring `"@embed" : "true"` in the frame I would expect that object URIs would be embedded in all fields declared as such. But they are only once embedded (arbitrarily in one of the declared fields) - if the reference URI appears in different fields at the same time the other fields only get the reference URI, nothing is embedded there.

There are solutions in different languages, but not in java yet. Would be glad to have it here, too!

See also https://github.com/digitalbazaar/jsonld.js/issues/20.
See also https://github.com/json-ld/json-ld.org/issues/377.

**Comment by boutros** [comment]
I'm also struggelig with this limitation.

I'm not up to date on the JSON-LD specifications, but would a pull-request be accepted if it implemented the following logic from https://github.com/json-ld/json-ld.org/issues/377 ?
 
> A new experimental embed API has been implemented in jsonld.js. Instead of using "@embed": true/false, three options for the @embed directive are available: @always, @never, and @last. The current framing behavior for "@embed": true matches that of @last which is the default. There's backwards compatibility support in jsonld.js for the previous @embed boolean flags.

**Comment by ansell** [comment]
If you have a pull request then I would review it. Basically, if there are no regressions and it is compatible with another client then it would be acceptable, given that the Framing part of the spec hasn't been standardised yet and could benefit from some innovation still.

**Comment by boutros** [comment]
Good.

I'm not sure what should be the appropriate way to handle the following scenario:

> @embed is set to @always
> node A references (and thus embeds) node B
> node B references (and thus embeds) node A

This would lead to infinite recurison. It is not possible to have a circular dependency in JSON, so in this case the prosessor must either throw an error, or fall back to using @last, for example.
I'll check how jsonld.js handles this

**Comment by RavinderSinghMaan** [comment]
I have implemented below component to fix the problem. In case it helps someone. It uses kryo for deep cloning. So you will need to add below as dependency. If I get some more time I will try to implement embed @always in the JsonLdProcessor.

```
            <dependency>
                <groupId>com.esotericsoftware</groupId>
                <artifactId>kryo</artifactId>
                <version>4.0.0</version>
            </dependency>

```


```package com.pearson.cms.converters;

import com.esotericsoftware.kryo.Kryo;
import org.slf4j.LoggerFactory;

import java.util.*;

public class JSONLDFrameExpander {
    private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(JSONLDFrameExpander.class);


    private Kryo kryo = new Kryo();

    Map expand(Map frame) {
        long currentTimeMillisToFix = System.currentTimeMillis();
        Map idToObjectMap = new HashMap();
        createIdToObjectMap(clone(frame), idToObjectMap);
        long endTimeToCreateMap = System.currentTimeMillis();
        if (LOG.isDebugEnabled()) {
            LOG.debug(String.format("Framing time to create IdTOObjectMap : %d",
                    endTimeToCreateMap - currentTimeMillisToFix
            ));
        }

        Map planeIDToObjectMap = (Map)clone(idToObjectMap);
        for(Object key: planeIDToObjectMap.keySet()) {
            Object o = planeIDToObjectMap.get(key);
            removeLinks(o, idToObjectMap);
            planeIDToObjectMap.put(key, o);
        }
        long endTimeToRemoveLinks = System.currentTimeMillis();

        if (LOG.isDebugEnabled()) {
            LOG.debug(String.format("Framing time to ToRemoveLinks : %d",
                    endTimeToRemoveLinks - endTimeToCreateMap
            ));
        }

        correctEmbededObjects(frame, planeIDToObjectMap, new HashSet<>());

        long endTimeToCorrectEmbededObjects = System.currentTimeMillis();
        if (LOG.isDebugEnabled()) {
            LOG.debug(String.format("Framing time to correctEmbededObjects : %d",
                    endTimeToCorrectEmbededObjects - endTimeToRemoveLinks
            ));
        }

        return frame;
    }

    void removeLinks(Object object, Map idToObjectMap) {
        if (object instanceof Map) {
            Map map = (Map)object;
            Set<Map.Entry> entrySet = map.entrySet();
            for(Map.Entry entry: entrySet) {
                Object value = entry.getValue();
                if (value instanceof Map) {
                    Map valueMap = (Map)value;
                    Object idValue = valueMap.get("id");
                    if(idValue != null && !idValue.toString().startsWith("_") && idToObjectMap.containsKey(idValue)) {
                        Map newMap = new LinkedHashMap();
                        newMap.put("id", idValue);
                        entry.setValue(newMap);
                    }
                } else if(value instanceof List) {
                    List list = (List) value;
                    int i = 0;
                    for(Object listObject:list) {
                        if (listObject instanceof Map) {
                            Map valueMap = (Map)listObject;
                            Object idValue = valueMap.get("id");
                            Map newMap = new LinkedHashMap();
                            newMap.put("id", idValue);
                            list.set(i, newMap);
                        }
                        i++;
                        //TODO if listObject is Array recurse
                    }
                }


            }
        }
    }

    void correctEmbededObjects(Object frame, Map idToObjectMap, Set<String> ancestors) {
        if(frame instanceof List) {
            int i = 0;
            for(Object object: (List) frame) {
                if (object instanceof String && idToObjectMap.containsKey(object) && !ancestors.contains(object)) {
                    Map clone = (Map) clone(idToObjectMap.get(object));
                    Set<String> ancestorsClone = addToAncesors(ancestors, object.toString());
                    ((List) frame).set(i , clone);
                    correctEmbededObjects(clone, idToObjectMap, ancestorsClone);
                } else {
                    correctEmbededObjects(object, idToObjectMap, ancestors);
                }
                i++;
            }

        } else if(frame instanceof Map) {
            Map frameMap = (Map) frame;
            Set<Map.Entry> entrySet = frameMap.entrySet();
            if(entrySet.size() == 1) {
                for (Map.Entry entry : entrySet) {
                    Object entryValue = entry.getValue();
                    if (entry.getKey().equals("id") && !ancestors.contains(entryValue)) {
                        Map clone = (Map) clone(idToObjectMap.get(entryValue));
                        frameMap.putAll(clone);

                        Set<String> ancestorClone = addToAncesors(ancestors, entryValue.toString());
                        correctEmbededObjects(clone, idToObjectMap, ancestorClone);
                    } else if (entryValueIsLiteralAndCanBeExpanded(idToObjectMap, ancestors, entryValue)) {
                        expandLiteral(idToObjectMap, ancestors, entry, entryValue);
                    } else {
                        correctEmbededObjects(entryValue, idToObjectMap, ancestors);
                    }
                }
            } else {
                for (Map.Entry entry : entrySet) {
                    Object key = entry.getKey();
                    Object entryValue = entry.getValue();
                    if (key.equals("id")) {
                        ancestors = addToAncesors(ancestors, entryValue.toString());
                    } else if (entryValueIsLiteralAndCanBeExpanded(idToObjectMap, ancestors, entryValue)) {
                        expandLiteral(idToObjectMap, ancestors, entry, entryValue);
                    } else {
                        correctEmbededObjects(entryValue, idToObjectMap, ancestors);
                    }
                }
            }
        }
    }

    private Set<String> addToAncesors(Set<String> ancestors, String newValue) {
        ancestors = (Set<String>) clone(ancestors);
        ancestors.add(newValue);
        return ancestors;
    }

    private void expandLiteral(Map idToObjectMap, Set<String> ancestors, Map.Entry entry, Object entryValue) {
        ancestors = addToAncesors(ancestors, entryValue.toString());
        Map clone = (Map) clone(idToObjectMap.get(entryValue));
        entry.setValue(clone);
        correctEmbededObjects(clone, idToObjectMap, ancestors);
    }

    private boolean entryValueIsLiteralAndCanBeExpanded(Map idToObjectMap, Set<String> ancestors, Object entryValue) {
        return entryValue instanceof String && !ancestors.contains(entryValue) && idToObjectMap.containsKey(entryValue);
    }

    void createIdToObjectMap(Object frame, Map idToObjectMap) {
        if(frame instanceof List) {
            for(Object object: (List) frame) {
                createIdToObjectMap(object, idToObjectMap);
            }
        } else if(frame instanceof Map) {
            Map map = (Map) frame;
            Set<Map.Entry> entrySet = map.entrySet();
            for (Map.Entry entry: entrySet) {
                Object entryValue = entry.getValue();
                if(keyIsIdAndKeyIsNotInIdToObjectMap(idToObjectMap, entry, entryValue)) {
                    idToObjectMap.put(entryValue, map);
                } else if (entryValueHasDescendantObjects(entryValue)) {
                    createIdToObjectMap(entryValue, idToObjectMap);
                }
            }
        }
    }

    private boolean entryValueHasDescendantObjects(Object entryValue) {
        return entryValue instanceof Map || entryValue instanceof List;
    }

    private boolean keyIsIdAndKeyIsNotInIdToObjectMap(Map idToObjectMap, Map.Entry entry, Object entryValue) {
        return entry.getKey().equals("id") && !idToObjectMap.containsKey(entryValue);
    }

    Object clone(Object value) {
        return kryo.copy(value);
    }


}

```

If you add above class into your src then after call to frame method from JsonLdProcessor you can call expand method. See example below
```
        Map frame = JsonLdProcessor.frame(content, frameWithContext, new JsonLdOptions(""));
        frame = new JSONLDFrameExpander.expand(frame);

```

**Comment by eroux** [comment]
I think this is now handled by the (recently introduced) handling of `"@embed": "@always"`, @dr0i, @boutros, @RavinderSinghMaan  can you confirm?

**Comment by RavinderSinghMaan** [comment]
@eroux embed always option is there in spec. But it is not implemented in this Java library. It is implemented in JavaScript library.

**Comment by eroux** [comment]
@RavinderSinghMaan I've handled `"@embed": "@always"` in https://github.com/jsonld-java/jsonld-java/pull/210/commits/4fe56fd4edcf943782879023483284ab8ad08c65 please test

**Comment by dr0i** [comment]
I wrote a junit test for it and fixed a small bug. It shows that the embedding of multiple objects now indeed works :)
Thank you @eroux!

Opened a PR against your master - have a look.

**Comment by RavinderSinghMaan** [comment]
Fantastic @eroux ! Well done. I did a quick test with a large input and result looks good.
One thing which was really putting me off from implementing it as per algorithm was performance. So I started exploring the alternative ways to implement same but with performance as the top priority. We use Java sesame library and results we get from triplestore is sesame Model. To serialise it to framed JSONLD output first it has to be converted to object which frame method expect and then frame it. All of it is taking too much time. For result of 20 resources, serialiasation itself with jsonld-java library is taking 70 to 90 millis. So what I have started implementing is to bypass those steps and serialise straight to JSONLD. This has helped a lot to imporve the performance it now takes 5-10 millis to frame 20 results with embed @always. Also it is very stripped down version of the spec just focusing on the use case. If you guys wish I can put it all on github and may be we can improve it further.
